schemas/fluidity_options.rnc

```diff
@@ -3718,6 +3787,107 @@ generic_particle_attribute_choice =
       }
    )
 
+generic_vector_particle_attribute_choice =
+   (
+      ## Constant value
+      element constant {
+         real_dim_vector
+      }|
+      ## Python function prescribing real input. Functions should be of the form:
+      ##
+      ##  def val(X, t, dt):
+      ##     # Function code
+      ##     return # Return value
+      ##
+      ## where X is a tuple of length geometry dimension.
+      element python {
+         python_code
+      }|
+      ## Python function prescribing real input. Functions should be of the form:
+      ##
+      ##  def val (X, t, dt, fields):
+      ##     # Function code
+      ##     return # Return value
+      ##
+      ##  where X is a tuple of length geometry dimension.
+      ##
+      ##  E.g. if Velocity field and positions vector are required
+      ##  def val (X, t, dt, fields):
+      ##    from numpy import zeros
+      ##    y = zeros(2,float)
+      ##    for i in range(0,2):
+      ##       y[i] = X[i] + fields["Velocity"][i]
+      ##    return y
+      element python_fields {
+         python_code,
+         ## Store attribute values from previous timestep on particles.
+         element store_old_attribute {
+            comment
+         }?
+      }|
+      ## File containing the particle positions in binary form
+      element from_checkpoint_file {
+         attribute file_name { xsd:string },
+         ## The format of the input file containing field data.
+         element format {
+            element string_value {
+               "binary"
+            }
+         }
+      }
+   )
+
+generic_tensor_particle_attribute_choice =
+   (
+      ## Constant value
+      element constant {
+         real_dim_tensor
+      }|
+      ## Python function prescribing real input. Functions should be of the form:
+      ##
+      ##  def val(X, t, dt):
+      ##     # Function code
+      ##     return # Return value
+      ##
+      ## where X is a tuple of length geometry dimension.
```

Comment:
And the return value is dim X dim tensor in the form of a list of lists (?) or anything compatible with a dim X dim numpy array I presume?

Reply:
Whoops, good catch! We ask numpy for `PyArray_DIMS` after `PyArray_ContiguousFromObject`, so I think your presumption is correct.